node/toolrunner.ts

```diff
@@ -584,27 +588,48 @@ export class ToolRunner extends events.EventEmitter {
                 this.pipeOutputToTool._getSpawnArgs(options),
                 this.pipeOutputToTool._getSpawnOptions(options));
 
+            let fileStream: fs.WriteStream = this.pipeOutputToFile ? fs.createWriteStream(this.pipeOutputToFile) : null;
+            if (fileStream) {
+                fileStream.write(this._getCommandString(options) + os.EOL);
+            }
+
             //pipe stdout of first tool to stdin of second tool
             cpFirst.stdout.on('data', (data: Buffer) => {
                 try {
+                    if (fileStream) {
+                        fileStream.write(data.toString());
+                    }
                     cp.stdin.write(data);
                 } catch (err) {
                     this._debug('Failed to pipe output of ' + toolPathFirst + ' to ' + toolPath);
                     this._debug(toolPath + ' might have exited due to errors prematurely. Verify the arguments passed are valid.');
                 }
             });
             cpFirst.stderr.on('data', (data: Buffer) => {
+                if (fileStream) {
+                    fileStream.write(data.toString());
+                }
                 successFirst = !options.failOnStdErr;
                 if (!options.silent) {
                     var s = options.failOnStdErr ? options.errStream : options.outStream;
                     s.write(data);
                 }
             });
             cpFirst.on('error', (err) => {
+                if (fileStream) {
+                    fileStream.on('finish', () => {  
+                        fileStream.close();
+                    });
+                }
                 cp.stdin.end();
                 defer.reject(new Error(toolPathFirst + ' failed. ' + err.message));
             });
             cpFirst.on('close', (code, signal) => {
+                if (fileStream) {
```

Comment:
I wonder if this is reliable. Do you know whether `close` means no further stdout/stderr data events are pending to be drained? I believe the node docs have good details about this, but I don't remember the details offhand.

Reply:
It should be as per https://nodejs.org/api/child_process.html#child_process_event_close - The 'close' event is emitted when the stdio streams of a child process have been closed. This is distinct from the 'exit' event, since multiple processes might share the same stdio streams.

Reply:
I wrote tests and found that there were some timing issues. I fixed the issues but the code looks ugly. Will sync up before doing any refactoring.